### PR TITLE
Add Shared Context Capability to Step Definitions

### PR DESCRIPTION
**Pull request by pplancq**
This PR introduces the ability to share context between `steps` and `stepDefinition` files during test execution. This is achieved by using the context to load the necessary information.

Here is an example of how this works:

In the `stepDefinition` function, we initialise a `new VendingMachine()` via `beforeEach`. This allows it to be added to the context, which can then be used directly in steps or even in other `stepDefinition` files.

```typescript
// specs/step-definitions/global-vending-machine-steps.ts
import { StepDefinitionsWithContext } from 'jest-cucumber';

import { VendingMachine } from '../src/vending-machine';

type Context = {
  vendingMachine: VendingMachine;
};

export const globalVendingMachineSteps: StepDefinitionsWithContext<Context> = ({ given, and, when, then, context }) => {
  beforeEach(() => {
    context.vendingMachine = new VendingMachine();
  });
  
  given(/^the vending machine has "(.*)" in stock$/, (itemName: string) => {
    context.vendingMachine.stockItem(itemName, 1);
  });

  and('I have inserted the correct amount of money', () => {
    context.vendingMachine.insertMoney(0.5);
  });

  when(/^I purchase "(.*)"$/, (itemName: string) => {
    context.vendingMachine.dispenseItem(itemName);
  });

  then(/^my "(.*)" should be dispensed$/, (itemName: string) => {
    const inventoryAmount = context.vendingMachine.items[itemName];
    expect(inventoryAmount).toBe(0);
  });
};
```
```typescript
// specs/step-definitions/expect-vending-machine-steps.ts
import { StepDefinitionsWithContext } from 'jest-cucumber';

type Context = {
  vendingMachine: VendingMachine;
};

export const expectVendingMachineSteps: StepDefinitionsWithContext<Context> = ({ given, and, when, then, context }) => {
  then(/^my "(.*)" should be dispensed$/, (itemName: string) => {
    const inventoryAmount = context.vendingMachine.items[itemName];
    expect(inventoryAmount).toBe(0);
  });
};

```

### CODE REVIEW

**Comment by bencompton** [comment]
TL;DR - I disagree with global context and global step matching, but since I've already added global step matching as an option based on popular demand, global context is needed as well.

So I've considered this in the past, but I personally never have been a fan of Cucumber's 'World' or any notion of global state or "context" because of the Interface Segregation Principal. Over time, you end up with this huge monstrosity that gets passed around to every test and it can become a dumping ground for every random concern that can eventually become difficult to understand and maintain.

Similarly, I'm not a fan of global step matching and making everything global by default. The stronger cohesion of Jest, Jasmine, RSpec, etc. style tests is a big part of the reason why I created this library as an alternative to Cucumber.js in the first place. In Jest tests, all tests for a particular concern reside in a single file or co-located set of files, and dependencies of each test are declared inline, typically initialized in 'beforeEach' blocks right next to the tests themselves. Basically, it's not necessary to jump all around in the codebase to understand how a test works like Cucumber requires. In most cases, you can read only the code in the test file and have a good understanding of how each test and feature it's testing works. The only thing I don't like about Jest is encouraging the use of trickery with mocks instead of writing clean, testable, loosely coupled code.

I went ahead and added 'autoBindSteps' anyway because a whole lot of people don't care about everything I said above and just want it to work like Cucumber. A lot of people are also using this package to drive UI interactions in E2E tests instead of for using it for its intended purpose of writing acceptance tests around domain rules. Others also may not have had the same experience as I have using Cucumber and ending up with a difficult to maintain codebase years later, or maybe they've found other techniques to manage the complexity as their codebase scaled.

Despite my efforts to find something better than global context to use with 'autoBindSteps', global step matching kind of necessitates something like this.

I don't mean to come off as rude in any way and am grateful for this contribution. For the sake of the community, I am happy to agree to disagree and accept this contribution. The code looks good from first glance, but I just need to pull it down and play around with it a bit to make sure I haven't missed anything before merging and publishing.

**Comment by pplancq** [comment]
Thank you for your comment.
Although I don't want to impose anything, I would like to express my preference for the `autoBindSteps` concept. 
This syntax is easier to write and allows easier reuse of steps, in my opinion. 
I've got into the habit of dividing my steps into coherent definition files. 
For example, I like to have a file for steps related to API mocks (fetch or axios), an element search file, an element interaction file, or user-related steps such as authentication and rights.
Over the course of my projects, I've developed a library of reusable steps (which is exclusive to my company) that can be used in other projects like this.

```Javascript
import { mockSteps, userSteps, fooSteps, barSteps } from '@private/steps-lib';

export const localSteps: StepDefinitions = ({ given, and, when, then }) => {
  //...
};

autoBindSteps([feature], [ mockSteps, userSteps, fooSteps, barSteps, localSteps ]);
```

I use this method as part of my ReactJS applications with testing-library, and often need to initialise some elements at the start of a test or scenario, such as the `render(<Component />)` or `userEvent.setup()`, which are required in several steps.
While I understand your concern, it is crucial for me to reset the context between each scenario. I haven't implemented it here as I've got into the habit of doing it myself in my tests via `afterEach`. However, if necessary, I can examine if this can be defined by default, either in this RP, or in another, based on complexity implementation.

Also, if necessary, we could make the context usable only on request. To explain this via the `autoBindSteps` initialisation factory, I could include an argument to enable or disable the context, which would look something like this:

```javascript
export const autoBindSteps = createAutoBindSteps(jestLike);
export const autoBindStepsWithContext = createAutoBindSteps(jestLike, true);`
```

Finally, include a warning in the documentation to inform users of the potential dangers of abusing context.

What do you think?

**Review by bencompton — APPROVED** [review]

**Comment by bencompton** [comment]
Appreciate the additional insights into how you use the package! Agreed, easier step re-use is one of the main arguments I’ve heard in favor of the `autoBindSteps` approach, and I’m certainly willing agree to disagree and support the DevEx that a lot of people want to see.

Your PR is backwards compatible, so no breaking changes is good. It’s up to you whether to make it more explicitly optional as you suggested—I’m fine with it as-is since context is probably desirable most of the time.

I’ve reviewed everything, and the only thing that stands out in your PR is that another section to `docs/AutomaticStepBinding.md` explaining how to use `context` would be helpful, especially since the example is re-using an existing feature file and the example alone might not stand out. If you want to add that in a separate PR, that's fine too.

**Comment by pplancq** [comment]
Hello,

Thank you for your feedback and for merging this PR. Yes, indeed, I wanted to make sure I didn't introduce any breaks, as long as it's possible, I'll do my best. As for the documentation, I've taken note of it and I'll try to take a moment to complete it.

**Comment by bencompton** [comment]
This is a pleasant surprise! I came here to comment that this PR was released, but `semantic-release` beat me to it. Pretty slick, and thanks again!

**Comment by pplancq** [comment]
Yes, `semantic-release` is a great release management tool. I'm very happy you like it.
